### PR TITLE
protectedts: do not update pts meta table in 24.1

### PR DESCRIPTION
**Pull request by jayshrivastava**
Note that after https://github.com/cockroachdb/cockroach/pull/118772 is merged, all "old-style" protected timestamps will be migrated to the
"new-style", meaning that the PTS cache and PTS meta table will not be required to read PTS
records.

The metadata above refers to metadata in the `system.protected_ts_meta` table. This table had a single
row which tracked metadata for all PTS records (ex. the total size of all the records, the number of records,
and `version`). The `version` was a monotonically increasing integer used to mark every change to the
`system.protected_ts_records` table, such as adding or removing records.

This change bifurcates the PTS storage system to have two 'modes': with and without metadata. PTS
storage with metadata was the norm before 24.1. For 24.1 onwards, we want to maintain PTS records
without metadata. The reason which the metadata updates are being removed is because of contention
on the  `system.protected_ts_meta` table when multiple jobs attempt to update it.

These two modes are necessary to ensure things work in a mixed version state. If a new node updates
a PTS record without metadata, a replica on an old node might not see it and the replica might perform gc.
For this reason, nodes should always update metadata until all nodes are on 24.1 and are able to read
all PTS records.

Removing this metadata has 2 main implications:
(a) Removing the size tracking effectively deprecates `kv.protectedts.max_bytes`.
This setting used to limit the total bytes used by the `system.protected_ts_records`. Because
PTS records used to protect a list of spans, the size of records was a concern. Now, they protect
descriptor ids, so they are much smaller - there is no need to limit the size of the table.

(b) Removing the `version` increments will stop the PTS cache from refreshing
This is okay because PTS updates will be picked up from span configs. As of https://github.com/cockroachdb/cockroach/pull/118772, no jobs will
use the old-style PTS records in 24.1. Also as of $118772, there will be no readers/consumers of the PTS cache.

This change also deprecates the metadata fields in protobufs and in the PTS storage API.

```
message Metadata {
   uint64 version = 1 [deprecated=true];
   uint64 num_records = 2 [deprecated=true];
   uint64 num_spans = 3 [deprecated=true];
   uint64 total_bytes = 4 [deprecated=true];
}
```

This change does not:
- remove the pts cache
- remove the pts meta table
- remove deprecated PTS creation (ie. `DisableProtectedTimestampForMultiTenant` testing knob which could have been removed in 22.2)
These can be removed in future commits as a part of the cleanup.

Informs: https://github.com/cockroachdb/cockroach/issues/82888

Release note: None

### CODE REVIEW

**Comment by jayshrivastava** [comment]
More context here 
https://cockroachlabs.slack.com/archives/C03JCUUSCD6/p1705947756397919?thread_ts=1705600153.906209&cid=C03JCUUSCD6

"""
Hey just an FYI, it looks like scheduled backups did not do PTS chaining until 22.2 (https://github.com/cockroachdb/cockroach/pull/68446). Deprecated PTS records should only exist today if they were created in 22.1 or earlier (https://github.com/cockroachdb/cockroach/issues/82888). Therefore, there shouldn't be any schedules running today which have old-style PTS records. Feel free to correct me :slightly_smiling_face:

Also, we didn't use PTS for streaming until 23.1 https://github.com/cockroachdb/cockroach/commit/721cc761de83e1424de2ec377c6078983ea11e4f
In summary,
changefeeds - we may have long running jobs with old-style pts records
backups/restores - jobs created < 22.2 should be completed
streaming - did not use pts until 23.1
scheduled backups - did not use pts until 22.2
"""

**Comment by jayshrivastava** [comment]
@rharding6373 Note that this PR is based on https://github.com/cockroachdb/cockroach/pull/118772, so only the last commit needs to be reviewed. I would appreciate it if you could take a look, but no rush at all :)

**Review by rharding6373 — COMMENTED** [review]
Nice work! I just have a few comments.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/112093#-:-NqVi4xyCSbLB6e287fp:bepnlze)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @jayshrivastava and @msbutler)

___
*[`pkg/kv/kvserver/protectedts/ptstorage/storage.go` line 99 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/112093#-NqVga3G367piskQUrOx:-NqVga3G367piskQUrOy:b-rtzg68) ([raw file](https://github.com/cockroachdb/cockroach/blob/641f894440e2d8daa36a35fa115b37d41728a45b/pkg/kv/kvserver/protectedts/ptstorage/storage.go#L99)):*
> ```Go
> 		len(r.DeprecatedSpans), encodedTarget, encodedTarget}
> 
> 	if !updateMeta {
> ```

Since updateMeta == true is the old, deprecated way, and we expect most of the time we'd be doing this !updateMeta path, I think we should switch these and put the updateMeta == false path as default and conditionally execute the updateMeta == true path. I.e., instead of what is here, do:

```
query = protectInsertRecordCTE
args = []interface{}{
			r.ID, r.Timestamp.AsOfSystemTime(),
			r.MetaType, meta,
			len(r.DeprecatedSpans), encodedTarget, encodedTarget}
if updateMeta {
        // Add some comment here about how this path updates the deprecated PTS meta table and why it's still here.
	query := protectQuery
	args := []interface{}{s.maxSpans, s.maxBytes, len(r.DeprecatedSpans),
		r.ID, r.Timestamp.AsOfSystemTime(),
		r.MetaType, meta,
		len(r.DeprecatedSpans), encodedTarget, encodedTarget}
}

```
___
*[`pkg/kv/kvserver/protectedts/ptstorage/storage_test.go` line 68 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/112093#-NqVch1c1zvIB2Avpsjm:-NqVch1c1zvIB2Avpsjn:b-zholw1) ([raw file](https://github.com/cockroachdb/cockroach/blob/641f894440e2d8daa36a35fa115b37d41728a45b/pkg/kv/kvserver/protectedts/ptstorage/storage_test.go#L68)):*
> ```Go
> 				// only test metadata. Run these tests once only. We don't
> 				// need to run them with `runWithMetaTable` = false.
> 				if withMetaTable && !test.runWithMetaTable {
> ```

I think this will still run some tests twice, just without the `withMetaTable` suffix. Maybe there should be a case like

```
if !withMetaTable && test.runWithMetaTable { 
  continue
}
```

and add a suffix in the case:
```
if withMetaTable || test.runWithMetaTable{
  name = ....
}
```
___
*[`pkg/kv/kvserver/protectedts/ptstorage/storage_test.go` line 657 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/112093#-NqVdS-y0KcmMV6_EH-1:-NqVdS-y0KcmMV6_EH-2:b-91e9tq) ([raw file](https://github.com/cockroachdb/cockroach/blob/641f894440e2d8daa36a35fa115b37d41728a45b/pkg/kv/kvserver/protectedts/ptstorage/storage_test.go#L657)):*
> ```Go
> 	}
> 
> 	testutils.RunTrueAndFalse(t, "using meta table", func(t *testing.T, withMetaTable bool) {
> ```

I was unfamiliar with this test util. TIL.
___
*[`pkg/spanconfig/spanconfigptsreader/adapter_test.go` line 50 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/112093#-NqVdj_5233YD6YXfmAk:-NqVdj_5233YD6YXfmAl:bvnsrmn) ([raw file](https://github.com/cockroachdb/cockroach/blob/641f894440e2d8daa36a35fa115b37d41728a45b/pkg/spanconfig/spanconfigptsreader/adapter_test.go#L50)):*
> ```Go
> 	ms := &manualSubscriber{}
> 
> 	// TODO: delete the cache in v24.2.
> ```

Could you create an issue to track this cleanup and put the issue number here? Thanks!
___
*[`pkg/spanconfig/spanconfigptsreader/adapter_test.go` line 112 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/112093#-NqVeUnr7DlGxWVVakqd:-NqVeUnr7DlGxWVVakqe:bfdud17) ([raw file](https://github.com/cockroachdb/cockroach/blob/641f894440e2d8daa36a35fa115b37d41728a45b/pkg/spanconfig/spanconfigptsreader/adapter_test.go#L112)):*
> ```Go
> 
> 		// Even though the cache has lower freshness, ignore it.
> 		mc.asOf = ts(10)
> ```

Seems like there's a lot of code copied from the test above. Could we reuse the code and just run the test with a bool to indicate with or without pts cache?

**Review by jayshrivastava — COMMENTED** [review]
TYFR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/112093#-:-Nqhqor4Bn7Z_zrMy3uy:b-80slv9)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @msbutler and @rharding6373)

___
*[`pkg/kv/kvserver/protectedts/ptstorage/storage.go` line 99 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/112093#-NqVga3G367piskQUrOx:-NqhqpwKD7twpG6FZB-Q:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/641f894440e2d8daa36a35fa115b37d41728a45b/pkg/kv/kvserver/protectedts/ptstorage/storage.go#L99)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

Since updateMeta == true is the old, deprecated way, and we expect most of the time we'd be doing this !updateMeta path, I think we should switch these and put the updateMeta == false path as default and conditionally execute the updateMeta == true path. I.e., instead of what is here, do:

```
query = protectInsertRecordCTE
args = []interface{}{
			r.ID, r.Timestamp.AsOfSystemTime(),
			r.MetaType, meta,
			len(r.DeprecatedSpans), encodedTarget, encodedTarget}
if updateMeta {
        // Add some comment here about how this path updates the deprecated PTS meta table and why it's still here.
	query := protectQuery
	args := []interface{}{s.maxSpans, s.maxBytes, len(r.DeprecatedSpans),
		r.ID, r.Timestamp.AsOfSystemTime(),
		r.MetaType, meta,
		len(r.DeprecatedSpans), encodedTarget, encodedTarget}
}

```
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/protectedts/ptstorage/storage_test.go` line 68 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/112093#-NqVch1c1zvIB2Avpsjm:-NqhqqT80tXNzfGhOely:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/641f894440e2d8daa36a35fa115b37d41728a45b/pkg/kv/kvserver/protectedts/ptstorage/storage_test.go#L68)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

I think this will still run some tests twice, just without the `withMetaTable` suffix. Maybe there should be a case like

```
if !withMetaTable && test.runWithMetaTable { 
  continue
}
```

and add a suffix in the case:
```
if withMetaTable || test.runWithMetaTable{
  name = ....
}
```
</blockquote></details>

Done.
___
*[`pkg/spanconfig/spanconfigptsreader/adapter_test.go` line 50 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/112093#-NqVdj_5233YD6YXfmAk:-NqhqrdK2sWK2IXZOy7L:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/641f894440e2d8daa36a35fa115b37d41728a45b/pkg/spanconfig/spanconfigptsreader/adapter_test.go#L50)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

Could you create an issue to track this cleanup and put the issue number here? Thanks!
</blockquote></details>

Done.
___
*[`pkg/spanconfig/spanconfigptsreader/adapter_test.go` line 112 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/112093#-NqVeUnr7DlGxWVVakqd:-NqhqsqaEbX0Uqu4lljR:b-7kqbjw) ([raw file](https://github.com/cockroachdb/cockroach/blob/641f894440e2d8daa36a35fa115b37d41728a45b/pkg/spanconfig/spanconfigptsreader/adapter_test.go#L112)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

Seems like there's a lot of code copied from the test above. Could we reuse the code and just run the test with a bool to indicate with or without pts cache?
</blockquote></details>

I felt like it would be easier to read with two copies - otherwise there would be if elses everywhere. The 2nd copy can be deleted pretty soon in 24.2.

**Review by rharding6373 — APPROVED** [review]
:lgtm: Great!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/112093#-:-NqiIws_EDwVLdd_r3g2:bioon7n)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @msbutler)

___
*[`pkg/spanconfig/spanconfigptsreader/adapter_test.go` line 50 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/112093#-NqVdj_5233YD6YXfmAk:-NqiHxVtDgG5GhipRJxP:b-ucfknh) ([raw file](https://github.com/cockroachdb/cockroach/blob/641f894440e2d8daa36a35fa115b37d41728a45b/pkg/spanconfig/spanconfigptsreader/adapter_test.go#L50)):*
<details><summary><i>Previously, jayshrivastava (Jayant) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Thanks!
___
*[`pkg/spanconfig/spanconfigptsreader/adapter_test.go` line 112 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/112093#-NqVeUnr7DlGxWVVakqd:-NqiI7TXEale_Sd8M_ih:b14z1gp) ([raw file](https://github.com/cockroachdb/cockroach/blob/641f894440e2d8daa36a35fa115b37d41728a45b/pkg/spanconfig/spanconfigptsreader/adapter_test.go#L112)):*
<details><summary><i>Previously, jayshrivastava (Jayant) wrote…</i></summary><blockquote>

I felt like it would be easier to read with two copies - otherwise there would be if elses everywhere. The 2nd copy can be deleted pretty soon in 24.2.
</blockquote></details>

Ah, at first glance they look very similar. I don't feel so strongly about eliminating the redundancy, so I'll leave it up to you.

**Comment by jayshrivastava** [comment]
bors r=rharding6373

Tyfr :)

**Comment by daniel-crlabs** [comment]
Customer is asking if this can be backported to `23.2`, they ran into this on `23.1.13` and will be upgrading to `23.2.10`

cc @rharding6373 and @jayshrivastava

**Comment by rharding6373** [comment]
This work (or rather the previous work that it depends on) can't be backported due to requiring a version gate, unfortunately. We can't backport it to minor versions.

**Comment by daniel-crlabs** [comment]
Sounds good, thank you for the response.
